project/VersionAxes.scala

```diff
@@ -0,0 +1,99 @@
+import sbt.Keys.*
+import sbt.internal.ProjectMatrix
+import JacocoSetup.jacocoSettings
+import com.github.sbt.jacoco.JacocoKeys.jacocoReportSettings
+
+
+object VersionAxes {
+
+  lazy val printVersionInfo = taskKey[Unit]("Print main dependency versions atum-service is being built for.")
+
+  case class SparkVersionAxis(sparkVersion: String) extends sbt.VirtualAxis.WeakAxis {
+
+    val sparkVersionMajor: String = Dependencies.Versions.getVersionUpToMajor(sparkVersion)
+    override val directorySuffix = s"-spark$sparkVersionMajor"
+    override val idSuffix: String = directorySuffix.replaceAll("""\W+""", "_")
+  }
+
+  private def camelCaseToLowerDashCase(origName: String): String = {
+    origName
+      .replaceAll("([A-Z])", "-$1")
+      .toLowerCase()
+  }
+
+  implicit class ProjectExtension(val projectMatrix: ProjectMatrix) extends AnyVal {
+
+    def sparkRow(sparkAxis: SparkVersionAxis,
+                 scalaVersions: Seq[String],
+                 dependenciesFnc: (String, String) => Seq[ModuleID],
+                 settings: Def.SettingsDefinition*): ProjectMatrix = {
+      val sparkVersion = sparkAxis.sparkVersion
+      scalaVersions.foldLeft(projectMatrix) { case (currentProjectMatrix, scalaVersion) =>
+        currentProjectMatrix.customRow(
+          scalaVersions = Seq(scalaVersion),
+          axisValues = Seq(sparkAxis, VirtualAxis.jvm),
+          _.settings(
+            moduleName := camelCaseToLowerDashCase(name.value + sparkAxis.directorySuffix),
+            scalacOptions ++= Setup.clientScalacOptions(scalaVersion),
+            libraryDependencies ++= dependenciesFnc(sparkVersion, scalaVersion),
+            printVersionInfo := streams.value.log.info(s"Building ${name.value} with Spark $sparkVersion, Scala $scalaVersion"),
+            (Compile / compile) := ((Compile / compile) dependsOn printVersionInfo).value,
+            jacocoReportSettings := jacocoSettings(sparkVersion, scalaVersion, name.value),
+          ).settings(settings *)
+        )
+      }
+    }
+
+    def scalasRow(scalaVersions: Seq[String],
+                  dependenciesFnc: String => Seq[ModuleID],
+                  settings: Def.SettingsDefinition*): ProjectMatrix = {
+      scalaVersions.foldLeft(projectMatrix)((currentProjectMatrix, scalaVersion) =>
+        currentProjectMatrix.customRow(
+          scalaVersions = Seq(scalaVersion),
+          axisValues = Seq(VirtualAxis.jvm),
+          _.settings(
+            scalacOptions ++= Setup.clientScalacOptions(scalaVersion),
+            libraryDependencies ++= dependenciesFnc(scalaVersion),
+            printVersionInfo := streams.value.log.info(s"Building ${name.value} with Scala $scalaVersion"),
+            (Compile / compile) := ((Compile / compile) dependsOn printVersionInfo).value,
+            jacocoReportSettings := jacocoSettings(scalaVersion, name.value),
+          ).settings(settings *)
+        )
+      )
+    }
+
+    def singleRow(scalaVersion: String,
```

Comment:
Another candidate for renaming as the method's name doesn't tell me anything about what it does.

Reply:
This is actually a brilliant idea. I was following the projectmatrix name from examples and how it is used in other examples. But I admit, it took me a while to grasp the concept. A better name would make it much easier to understand 👍